### PR TITLE
Fixed #34345 -- Added system check for ManyToManyFields with intermediate tables in ModelAdmin.filter_horizontal/vertical.

### PR DESCRIPTION
**Pull request by hrushikeshrv**
Started work on patch. Yet to add documentation and tests.

### CODE REVIEW

File: tests/modeladmin/test_checks.py [hunk 1]
```diff
@@ -321,6 +328,18 @@ class TestModelAdmin(ModelAdmin):
             "admin.E020",
         )
 
+    def test_invalid_m2m_field_with_through(self):
+        class TestModelAdmin(ModelAdmin):
+            filter_vertical = ("bands",)
```
Comment (felixxm) [inline tests/modeladmin/test_checks.py:333]:
```suggestion
            filter_vertical = ["bands"]
```

File: tests/modeladmin/test_checks.py [hunk 2]
```diff
@@ -363,6 +382,18 @@ class TestModelAdmin(ModelAdmin):
             "admin.E020",
         )
 
+    def test_invalid_m2m_field_with_through(self):
+        class TestModelAdmin(ModelAdmin):
+            filter_horizontal = ("bands",)
```
Comment (felixxm) [inline tests/modeladmin/test_checks.py:387]:
```suggestion
            filter_horizontal = ["bands"]
```

File: tests/modeladmin/models.py [hunk 3]
```diff
@@ -14,6 +19,13 @@ def __str__(self):
         return self.name
 
 
+class BandAssignment(models.Model):
+    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)
+    band = models.ForeignKey(Band, on_delete=models.CASCADE)
+    start_date = models.DateField()
+    end_date = models.DateField(null=True, blank=True)
```
Comment (felixxm) [inline tests/modeladmin/models.py:26]: I would move new models to test methods and remove unused columns, e.g.
```diff
diff --git a/tests/modeladmin/models.py b/tests/modeladmin/models.py
index a83f24c98c..51ebca768c 100644
--- a/tests/modeladmin/models.py
+++ b/tests/modeladmin/models.py
@@ -2,11 +2,6 @@ from django.contrib.auth.models import User
 from django.db import models
 
 
-class Artist(models.Model):
-    name = models.CharField(max_length=100)
-    bands = models.ManyToManyField("Band", through="BandAssignment")
-
-
 class Band(models.Model):
     name = models.CharField(max_length=100)
     bio = models.TextField()
@@ -19,13 +14,6 @@ class Band(models.Model):
         return self.name
 
 
-class BandAssignment(models.Model):
-    artist = models.ForeignKey(Artist, on_delete=models.CASCADE)
-    band = models.ForeignKey(Band, on_delete=models.CASCADE)
-    start_date = models.DateField()
-    end_date = models.DateField(null=True, blank=True)
-
-
 class Song(models.Model):
     name = models.CharField(max_length=100)
     band = models.ForeignKey(Band, models.CASCADE)
diff --git a/tests/modeladmin/test_checks.py b/tests/modeladmin/test_checks.py
index e2ce3e0f2e..ec18a309e3 100644
--- a/tests/modeladmin/test_checks.py
+++ b/tests/modeladmin/test_checks.py
@@ -4,13 +4,12 @@ from django.contrib.admin import BooleanFieldListFilter, SimpleListFilter
 from django.contrib.admin.options import VERTICAL, ModelAdmin, TabularInline
 from django.contrib.admin.sites import AdminSite
 from django.core.checks import Error
-from django.db.models import CASCADE, F, Field, ForeignKey, Model
+from django.db.models import CASCADE, F, Field, ForeignKey, ManyToManyField, Model
 from django.db.models.functions import Upper
 from django.forms.models import BaseModelFormSet
 from django.test import SimpleTestCase
 
 from .models import (
-    Artist,
     Band,
     Song,
     User,
@@ -329,8 +328,15 @@ class FilterVerticalCheckTests(CheckTestCase):
         )
 
     def test_invalid_m2m_field_with_through(self):
+        class Artist(Model):
+            bands = ManyToManyField("Band", through="BandArtist")
+
+        class BandArtist(Model):
+            artist = ForeignKey(Artist, on_delete=CASCADE)
+            band = ForeignKey(Band, on_delete=CASCADE)
+
         class TestModelAdmin(ModelAdmin):
-            filter_vertical = ("bands",)
+            filter_vertical = ["bands"]
 
         self.assertIsInvalid(
             TestModelAdmin,
@@ -383,8 +389,15 @@ class FilterHorizontalCheckTests(CheckTestCase):
         )
 
     def test_invalid_m2m_field_with_through(self):
+        class Artist(Model):
+            bands = ManyToManyField("Band", through="BandArtist")
+
+        class BandArtist(Model):
+            artist = ForeignKey(Artist, on_delete=CASCADE)
+            band = ForeignKey(Band, on_delete=CASCADE)
+
         class TestModelAdmin(ModelAdmin):
-            filter_horizontal = ("bands",)
+            filter_horizontal = ["bands"]
 
         self.assertIsInvalid(
             TestModelAdmin,
```

**Review by felixxm — COMMENTED** [review]
@hrushikeshrv Thanks :+1: I left small comments.

File: tests/modeladmin/models.py (hunk 3 above)
Comment (hrushikeshrv) [reply tests/modeladmin/models.py]: It seems defining the `Artist` model twice is leading to some test failures that I couldn't recreate on my machine.

Comment (felixxm) [reply tests/modeladmin/models.py]: `@isolate_apps` was missing.

**Pull request merged**